<h1>Assignment 1</h1>

<p>In this first assignment we are focusing on a classification problem.<br>
Our goal will be to use both the <b>K-Nearest Neighbors</b> and the <b>Decision Tree</b> techniques and try to compare the predictive models obtained.</p>

<h3>The Dataset</h3>
<p>For this assignment we will be using the sonar_train.csv data set</p>
<p>This data set contains signals obtained from a variety of different aspect angles, spanning 90 degrees for mines and 180 degrees for rocks. Each sample is a set of 60 numbers in the range 0.0 to 1.0, where each number represents the energy within a particular frequency band (integrated over a certain period of time). The output target contains the letter R if the object is a rock and M if it is a mine (metal cylinder).</p>

In [1]:
import numpy as np
import matplotlib.pyplot as plt
import pandas as pd
%matplotlib inline

In [4]:
#Opening the sonar_train.csv file
data = pd.read_csv('sonar.csv')

###Display informations and the different column/features names

In [6]:
data.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 208 entries, 0 to 207
Data columns (total 61 columns):
Band1     208 non-null float64
Band2     208 non-null float64
Band3     208 non-null float64
Band4     208 non-null float64
Band5     208 non-null float64
Band6     208 non-null float64
Band7     208 non-null float64
Band8     208 non-null float64
Band9     208 non-null float64
Band10    208 non-null float64
Band11    208 non-null float64
Band12    208 non-null float64
Band13    208 non-null float64
Band14    208 non-null float64
Band15    208 non-null float64
Band16    208 non-null float64
Band17    208 non-null float64
Band18    208 non-null float64
Band19    208 non-null float64
Band20    208 non-null float64
Band21    208 non-null float64
Band22    208 non-null float64
Band23    208 non-null float64
Band24    208 non-null float64
Band25    208 non-null float64
Band26    208 non-null float64
Band27    208 non-null float64
Band28    208 non-null float64
Band29    208 non-null floa

In [7]:
np.shape(data)

(208, 61)

<p>In this data set each band is a feature and the target is the type of the object</p>
<h4> <u>QUESTION 1 :</u> How many samples and how many features do we have?</h4>
<p><i>Type your answer here: </i></p>
<h4><u></u> We have 208 samples and each sample has 60 features.</h4>

In [8]:
###using the train_test_split command split the data set in two : 'train' and 'test' with a ratio ~75%-25%
from sklearn.cross_validation import train_test_split
train, test = train_test_split(data, train_size=0.75)
np.shape(train)

(156, 61)

In [9]:
np.shape(test)

(52, 61)

In [10]:
#data.Type

In [13]:
#The last column in our sets correspond to the target
###Count how many of 'R' and 'M' you have in total
R = data[data.Type=='R']
M = data[data.Type=='M']
print np.shape(R), ', There is 97 letters R'
#R

(97, 61) , There is 97 letters R


In [15]:
print np.shape(M), ', There is also 111 letters M'
#M

(111, 61) , There is also 111 letters M


In [16]:
#In order to make targets more easily usable we will replace 'R' and 'M' by 0 and 1 respectively.
###Do it for both sets using the command : .replace()
train_01 = train.replace(['M','R'],[1,0])
test_01 = test.replace(['M','R'],[1,0])
test_01.Type 
train_01_fin = train.drop('Type', axis = 1)
test_01_fin = test.drop('Type', axis = 1)

<h3>K-Nearest Neighbors approach</h3>

<p>In this section we will apply the <b>K-Nearest Neighbors</b> technique to the data set</p>

In [21]:
from sklearn.neighbors import KNeighborsClassifier
###Fille the brackets to use the brute force approach and 7 neighbors
knn = KNeighborsClassifier(algorithm='brute',n_neighbors=7)
#################################################################
### Train the Nearest Neighbor Classifier "knn" on the training set
knn.fit(train_01_fin,train_01.Type)

KNeighborsClassifier(algorithm='brute', leaf_size=30, metric='minkowski',
           metric_params=None, n_jobs=1, n_neighbors=7, p=2,
           weights='uniform')

<h4> <u>QUESTION 2 :</u> Which kind of impact has the choice of the algorithm?</h4>
<p><i>Type your answer here: </i></p>
<h4> Imroving the accuracy </h4>

In [22]:
#At any time it is possible to recall the parameters using :
knn.get_params()

{'algorithm': 'brute',
 'leaf_size': 30,
 'metric': 'minkowski',
 'metric_params': None,
 'n_jobs': 1,
 'n_neighbors': 7,
 'p': 2,
 'weights': 'uniform'}

<p>We now want to test the accuracy of the nearest neighbor classifier</p>


In [23]:
###Compute the target prediction for the elements of the test set and compare with their true value
predict = knn.predict(test_01_fin)
print predict

[1 0 1 1 0 0 0 0 0 0 0 0 1 1 1 1 1 1 0 1 1 0 1 1 0 1 1 1 1 1 1 0 1 1 0 1 1
 1 1 1 0 1 0 0 1 0 1 0 0 1 0 1]


In [24]:
#scikit actually includes an accuracy evaluation function :
#Compute the accuracy using knn.score()
accuracy = knn.score(test_01_fin,test_01.Type)
print accuracy

0.807692307692


<h4> <u>QUESTION 3 :</u> Is the accuracy acceptable?</h4>
<p><i>Type your answer here</i></p>
<h4>Yes, it is acceptable </h4>


<p>In the following cell repeat the training and the test steps with different knn parameters and try to find a good combination.</p>

In [25]:
knn = KNeighborsClassifier(algorithm='auto',n_neighbors=7,leaf_size=25,p=1,weights='distance')
knn.fit(train_01_fin,train_01.Type)
predict = knn.predict(test_01_fin)
accuracy = knn.score(test_01_fin,test_01.Type)
print accuracy

0.846153846154


<h4> <u>QUESTION 4 :</u> What is the best accuracy you get? What can you conclude?</h4>
<p><i>Type your answer here</i></p>
<p><i><h4>The accuracy is improved </h4> </i></p>

<h3>Decision Tree</h3>

<p>Using the same sets we now investigate the efficiency of a <b>decision tree classifier<b> </p>

In [26]:
from sklearn import tree
dtr = tree.DecisionTreeClassifier()
# train the decision tree classifier "dtr" <<<max_depth=5,max_leaf_nodes=30,splitter='best'
dtr.fit(train_01_fin,train_01.Type)

DecisionTreeClassifier(class_weight=None, criterion='gini', max_depth=None,
            max_features=None, max_leaf_nodes=None, min_samples_leaf=1,
            min_samples_split=2, min_weight_fraction_leaf=0.0,
            presort=False, random_state=None, splitter='best')

In [27]:
# Same as previously we can display some informations
dtr.get_params()

{'class_weight': None,
 'criterion': 'gini',
 'max_depth': None,
 'max_features': None,
 'max_leaf_nodes': None,
 'min_samples_leaf': 1,
 'min_samples_split': 2,
 'min_weight_fraction_leaf': 0.0,
 'presort': False,
 'random_state': None,
 'splitter': 'best'}

<h4> <u>QUESTION 5 :</u> From the list above, give the keywords influencing the test performed at a decison node?</h4>
<p><i>Type your answer here</i></p>
<p><i>The keywords influencing the test performed at a decison node are:<h4> 'max_leaf_nodes', 'splitter', 'min_samples_split' and 'max_depth'  </h4></i></p>
<p>In the follwing, using the work done previously test the accuracy of the decision tree classifier.</p>

In [33]:
predict1 = dtr.predict(test_01_fin)
####################################
accu = dtr.score(test_01_fin,test_01.Type)
print accu

0.788461538462


<p>Hereafter, using different parameters for our decision tree, try to improve the accuracy</p>

<h4> <u>QUESTION 6 :</u> Explain which parameter(s) you plan to change and why?</h4>
<p><i>Type your answer here:</i></p>
<h4> min_samples_split, it impacts in the accuracy.</h4>

In [37]:
dtr = tree.DecisionTreeClassifier(min_samples_split=1,max_depth=5,max_leaf_nodes=30,splitter='best')
dtr.fit(train_01_fin,train_01.Type)
predict1 = dtr.predict(test_01_fin)
accu = dtr.score(test_01_fin,test_01.Type)
print accu

0.826923076923


<p>The results are different but obviously when we identify mines we don't want to be wrong.</p>

<h4> <u>QUESTION 7 :</u> Which one of the two methods is better? Is the accuracy acceptable? Why?</h4>
<p><i>Type your answer here :</i></p> 
<h4>Decision Tree, yes its accuracy better.   </h4> 

<h4> <u>QUESTION 8 :</u> What can you conclude from the whole study</h4>
<p><i>Type your answer here:  </i></p>
<h4>The study shows that Decision Tree has more precise accuracy  than KNN</h4>

<h4> <u>BONUS :</u> Copy/Paste the code to generate a pdf file containing the best tree and attach the file generated to your assignment when sending everything to me</h4>
<p> Remark:</p>
<h4> Parser is not working until now, it will work in your computer. </h4>

In [38]:
from sklearn.externals.six import StringIO
import pydot

In [41]:
dot_data = StringIO()
f = tree.export_graphviz(dtr,out_file = dot_data,
                         filled=True,rounded=True)
graph = pydot.graph_from_dot_data(dot_data.getvalue())
graph.write_pdf('sonar.pdf')